jbpm-task-assigning/kie-server-client-task-assigning/src/main/java/org/kie/server/client/util/TaskDataReader.java

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright 2020 Red Hat, Inc. and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.kie.server.client.util;
+
+import java.time.LocalDateTime;
+import java.util.ArrayList;
+import java.util.List;
+
+import org.kie.server.api.model.taskassigning.TaskData;
+import org.kie.server.api.model.taskassigning.TaskDataList;
+import org.kie.server.api.model.taskassigning.TaskInputVariablesReadMode;
+import org.kie.server.client.TaskAssigningRuntimeClient;
+
+public class TaskDataReader {
+
+    private TaskAssigningRuntimeClient runtimeClient;
+
+    private TaskDataReader(TaskAssigningRuntimeClient runtimeClient) {
+        this.runtimeClient = runtimeClient;
+    }
+
+    public static class Result {
+
+        LocalDateTime queryTime;
+
+        List<TaskData> tasks;
+
+        private Result(LocalDateTime queryTime, List<TaskData> tasks) {
+            this.queryTime = queryTime;
+            this.tasks = tasks;
+        }
+
+        public LocalDateTime getQueryTime() {
+            return queryTime;
+        }
+
+        public List<TaskData> getTasks() {
+            return tasks;
+        }
+    }
+
+    public static TaskDataReader from(TaskAssigningRuntimeClient runtimeClient) {
+        return new TaskDataReader(runtimeClient);
+    }
+
+    public Result readTasks(long fromTaskId, List<String> status, LocalDateTime fromLastModificationDate,
+                            int pageSize, TaskInputVariablesReadMode taskInputVariablesReadMode) {
+        final List<TaskData> result = new ArrayList<>();
+        boolean finished = false;
+        List<TaskData> partialResult;
+        TaskDataList taskDataList;
+        TaskData lastItem = null;
+        LocalDateTime queryTime = null;
+
+        long taskId = fromTaskId;
+        int nextPageSize = pageSize;
+        while (!finished) {
+            taskDataList = runtimeClient.findTasks(taskId, status, fromLastModificationDate,
+                                                   0, nextPageSize, taskInputVariablesReadMode);
+            partialResult = new ArrayList<>(taskDataList.getItems());
```

Comment:
Would you please explain (possibly in comments) the logic of this loop?

Reply:
yes, in general the kie-server queries related APIs that iterates server side datasets are based on a paging reading. By passing the pageNumber and pageSize.
But here I need to read all the values, but, doing a paging reading.
So this loop implements "read all values" by getting chunks in a paged reading fashion.
Note the for feeding the solver e.g. for creating the initial solution, we need to consume all the tasks, and not only a page.